src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        Node<T> oldLast = last;
+        last = new Node<>(null, value, oldLast);
+        if (oldLast == null) {
+            first = last;
+        } else {
+            oldLast.next = last;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        outOfBoundsCheckToAdd(index);
+        Node<T> nodeToAdd = new Node<>(null, value, null);
+        Node<T> nodeNext = findNodeByIndex(index);
+        if (index == 0) {
+            first = nodeToAdd;
+            if (nodeNext != null) {
+                nodeToAdd.next = nodeNext;
+                nodeNext.prev = nodeToAdd;
+            } else {
+                last = first;
+            }
+            size++;
+        } else if (index == size) {
+            add(value);
+        } else {
+            nodeToAdd.prev = nodeNext.prev;
+            nodeToAdd.next = nodeNext;
+            Node<T> nodePrev = nodeNext.prev;
+            nodeNext.prev = nodeToAdd;
+            nodePrev.next = nodeToAdd;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
+        Node<T> nodeToSet = findNodeByIndex(index);
+        T oldNodeValue = nodeToSet.value;
+        nodeToSet.value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheckToGetSetRemove(index);
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        int currentIndex = 0;
+        while (currentIndex != size) {
+            if (currentNode.value == object
+                    || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
```

Comment:
```suggestion
        for (Node<T> x = first; x != null; x = x.next) {
            if (object == x.value || object != null && object.equals(x.value)) {
                unlink(x);
                return true;
            }
        }
```

Reply:
```suggestion
        Node<T> x = first; 
        while (x != null) {
            if (object == x.value || (object != null && object.equals(x.value))) {
                unlink(x);
                return true;
            }
            x = x.next;
        }
```
you don't need currentIndex